### DISCUSSION TITLE
Agregar atributo data-* a los diferentes elementos para poder acceder a ellos en pruebas automatizadas

### DISCUSSION THREAD

**Discussion by JaviOCC** [occmundial/occ-atomic#306] [Ideas]
Hola, estamos llevando acabo un proyecto de pruebas automatizadas para los flujos críticos de hirers y como parte de las buenas prácticas, se recomienda agregar un atributo a los elementos para que sean fáciles de acceder y evitar utilizar selectores que cambien frecuentemente.
Que les parece la idea de pasarle una prop a los componentes ej: **testId** y atomic agregue un atributo data-* ej: **data-testid** y así evitar futuros problemas en caso de que llegara a cambiar el id o clase de los diferentes elementos? Este atributo seria específicamente para pruebas.
Cabe mencionar que este cambio no afectaría a los sitios que actualmente hagan uso de la librería.

Ejemplo de uso:
![Captura2](https://user-images.githubusercontent.com/92745995/198107851-8cabb6e1-4f47-4657-a350-df9accf82638.PNG)

Y atomic lo agregue así:
![Captura3](https://user-images.githubusercontent.com/92745995/198108426-d11ceda6-e1b3-41dc-85ab-40a159192b43.PNG)

Estas son algunas herramientas que nos lo recomiendan como buenas practicas: 
[https://docs.cypress.io/guides/references/best-practices#Selecting-Elements](url)
[https://playwright.dev/docs/selectors#best-practices](url)

**Comment by Andrea-Esquivel** [comment]
@gpimentel91 @andr05 @rrosaG @MitziMtz @anakarencm @malagon91 @bloodysummers @Yasbek

**Comment by Drkger** [comment]
Yo veo bien el agregar estos data-* ya que tambien se pueden usar para tracking y cada vez veo mas que son usados en buenas practicas cuando los selectores o clases son dinamicos.

Aqui en la documentacion de Heap tambien recomiendan usarlos para poder tener un identificador unico e inmutable
https://help.heap.io/data-management/react-dom-minification/general-strategies-for-using-heap-with-a-minified-dom/

**Comment by jcuevas-occ-com-mx** [comment]
Si son más fáciles de acceder y evitamos usar selectores me parece una buena idea 👍

**Comment by cinthya-trejo-occ-com-mx** [comment]
Yo lo veo bien chicos 👍 solo hay que definir bien como va a quedar el atributo data-*

**Comment by cinthya-trejo-occ-com-mx** [comment]
y otra duda, en todos sus proyectos usan occ-atomic? o también https://github.com/occmundial/atomic ?

**Reply by JaviOCC** [reply]
Me parece que de lado de hirers no se utiliza esa librería. @gpimentel91

**Reply by Andrea-Esquivel** [reply]
En el caso del repo recruiter-selfService sí se hace uso de la librería que comenta @Yasbek

**Comment by ldiaz-occ-com-mx** [comment]
Ok, de acuerdo a la documentación que compartes sí parece la práctica correcta y recomendada apuntar a través de `data-test-id`. Como comenta Yas, entonces hay qué definir cómo vamos a estar manejando este atributo:
* Cómo se setea en un componente? (cómo debería enviarse el prop)
* Qué tipo de componentes lo requieren? (sólo componentes interactivos? componentes de layout también o de texto?)
* Puede un componente requerir más de un `data-test-id`?
  * Por ejemplo, en el componente Autocomplete debería poder configurarse un `data-test-id` para el input y otro para el droplist o algún otro subelemento? y cuál sería la convención para el nombrado de estos múltiples data-test-id en los props?
* Deberíamos definir alguna estrategia de nombrado para evitar _name collisions_? por ejemplo algo similar a la metodología [BEM](https://en.bem.info/methodology/naming-convention/) para que los data-test-id sean siempre únicos pero deterministas.

Esos son los puntos que se me ocurren ahorita, no sé si hay alguna otra cuestión que haya qué definir.

**Comment by JaviOCC** [comment]
Platicando mas de cerca con @ldiaz-occ, se propone lo siguiente
- De momento los componentes que se requiere tener este atributo serian los interactivos.
- Se les pasara la prop de testId la cual será opcional
- Para los componentes mas complejos ej: TextField, Modal, Droplist, se propone construir los ids de lado de atomic con base al testId que se le pase, como ejemplo en el Modal, pudieran existir tres componentes con los que se requiera interactuar: close icon, mainBtn, secBtn y la propuesta de nombrado tomando como base el testId = 'modal-datos-fiscales': 
   close icon = `${testId}_close-icon` = `modal-datos-fiscales_close-icon`
   mainBtn = `${testId}_main-button` = `modal-datos-fiscales_main-button`
   secBtn = `${testId}_secondary-button` = `modal-datos-fiscales_secondary-button`

Que les parece? @Andrea-Esquivel @gpimentel91 @ldiaz-occ @Yasbek @malagon91 @anakarencm @andr05 @DaniCuevas
